File: pilot/src/api/generators.rs
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::IntoResponse;
use axum::Json;
use serde::{Deserialize, Serialize};

use crate::app_state::AppState;
use super::error::ApiError;

#[derive(Debug, Deserialize)]
pub struct CreateGeneratorRequest {
    pub name: String,
    #[serde(default)]
    pub description: String,
    pub workload: std::collections::HashMap<String, f64>,
}

#[derive(Debug, Serialize)]
pub struct GeneratorResponse {
    pub name: String,
    pub description: String,
    pub workload: serde_json::Value,
    pub builtin: bool,
    pub created_at: String,
}

pub async fn list(
    State(state): State<AppState>,
) -> Result<impl IntoResponse, ApiError> {
    let generators = state.sqlite.list_generators().await?;
    let items: Vec<GeneratorResponse> = generators.iter().map(to_response).collect();
    Ok(Json(serde_json::json!({ "items": items })))
}

pub async fn get_one(
    State(state): State<AppState>,
    Path(name): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    let record = state
        .sqlite
        .get_generator(&name)
        .await?
        .ok_or_else(|| ApiError::NotFound(format!("generator '{name}' not found")))?;
    Ok(Json(to_response(&record)))
}

pub async fn create(
    State(state): State<AppState>,
    Json(req): Json<CreateGeneratorRequest>,
) -> Result<impl IntoResponse, ApiError> {
    if state.sqlite.get_generator(&req.name).await?.is_some() {
        return Err(ApiError::Conflict(format!("generator '{}' already exists", req.name)));
    }

    let workload_json = serde_json::to_string(&req.workload)
        .map_err(|e| ApiError::BadRequest(e.to_string()))?;

    state
        .sqlite
        .upsert_generator(&req.name, &req.description, &workload_json, false)
        .await?;

    let record = state
        .sqlite
        .get_generator(&req.name)
        .await?
        .ok_or_else(|| ApiError::Internal(anyhow::anyhow!("failed to read just-created generator")))?;

    Ok((StatusCode::CREATED, Json(to_response(&record))))
}

pub async fn delete(
    State(state): State<AppState>,
    Path(name): Path<String>,
) -> Result<impl IntoResponse, ApiError> {
    let record = state
        .sqlite
        .get_generator(&name)
        .await?
        .ok_or_else(|| ApiError::NotFound(format!("generator '{name}' not found")))?;

    if record.builtin != 0 {
        return Err(ApiError::Conflict("cannot delete built-in generator".to_string()));
    }

    state.sqlite.delete_generator(&name).await?;
    Ok(StatusCode::NO_CONTENT)
}

fn to_response(g: &crate::storage::sqlite::GeneratorRecord) -> GeneratorResponse {
    GeneratorResponse {
        name: g.name.clone(),
        description: g.description.clone(),
        workload: serde_json::from_str(&g.workload).unwrap_or_default(),
        builtin: g.builtin != 0,
        created_at: g.created_at.clone(),
    }
}

File: pilot/api/generators.go
package api

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/regret-io/regret/pilot-go/storage"
)

type generatorHandlers struct {
	state *AppState
}

// List handles GET /api/generators.
func (g *generatorHandlers) List(w http.ResponseWriter, r *http.Request) {
	generators, err := g.state.Sqlite.ListGenerators(r.Context())
	if err != nil {
		WriteError(w, InternalError(err))
		return
	}

	items := make([]GeneratorResponse, 0, len(generators))
	for i := range generators {
		items = append(items, toGeneratorResponse(&generators[i]))
	}
	WriteJSON(w, http.StatusOK, map[string]interface{}{"items": items})
}

// GetOne handles GET /api/generators/{name}.
func (g *generatorHandlers) GetOne(w http.ResponseWriter, r *http.Request) {
	name := chi.URLParam(r, "name")
	record, err := g.state.Sqlite.GetGenerator(r.Context(), name)
	if err != nil || record == nil {
		WriteError(w, NotFound(fmt.Sprintf("generator '%s' not found", name)))
		return
	}
	WriteJSON(w, http.StatusOK, toGeneratorResponse(record))
}

// Create handles POST /api/generators.
func (g *generatorHandlers) Create(w http.ResponseWriter, r *http.Request) {
	var req CreateGeneratorRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		WriteError(w, BadRequest("invalid JSON: "+err.Error()))
		return
	}

	ctx := r.Context()

	existing, _ := g.state.Sqlite.GetGenerator(ctx, req.Name)
	if existing != nil {
		WriteError(w, Conflict(fmt.Sprintf("generator '%s' already exists", req.Name)))
		return
	}

	workloadJSON, err := json.Marshal(req.Workload)
	if err != nil {
		WriteError(w, BadRequest(err.Error()))
		return
	}

	if err := g.state.Sqlite.UpsertGenerator(ctx, req.Name, req.Description, string(workloadJSON), int64(req.Rate), 0); err != nil {
		WriteError(w, InternalError(err))
		return
	}

	record, err := g.state.Sqlite.GetGenerator(ctx, req.Name)
	if err != nil || record == nil {
		WriteError(w, InternalError(fmt.Errorf("failed to read just-created generator")))
		return
	}

	WriteJSON(w, http.StatusCreated, toGeneratorResponse(record))
}

// Update handles PUT /api/generators/{name}.
func (g *generatorHandlers) Update(w http.ResponseWriter, r *http.Request) {
	name := chi.URLParam(r, "name")
	ctx := r.Context()

	existing, err := g.state.Sqlite.GetGenerator(ctx, name)
	if err != nil || existing == nil {
		WriteError(w, NotFound(fmt.Sprintf("generator '%s' not found", name)))
		return
	}

	var req CreateGeneratorRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		WriteError(w, BadRequest("invalid JSON: "+err.Error()))
		return
	}

	workloadJSON, err := json.Marshal(req.Workload)
	if err != nil {
		WriteError(w, BadRequest(err.Error()))
		return
	}

	builtin := 0
	if existing.Builtin != 0 {
		builtin = 1
	}

	if err := g.state.Sqlite.UpsertGenerator(ctx, name, req.Description, string(workloadJSON), int64(req.Rate), builtin); err != nil {
		WriteError(w, InternalError(err))
		return
	}

	record, err := g.state.Sqlite.GetGenerator(ctx, name)
	if err != nil || record == nil {
		WriteError(w, InternalError(fmt.Errorf("failed to read updated generator")))
		return
	}

	WriteJSON(w, http.StatusOK, toGeneratorResponse(record))
}

// Delete handles DELETE /api/generators/{name}.
func (g *generatorHandlers) Delete(w http.ResponseWriter, r *http.Request) {
	name := chi.URLParam(r, "name")
	ctx := r.Context()

	record, err := g.state.Sqlite.GetGenerator(ctx, name)
	if err != nil || record == nil {
		WriteError(w, NotFound(fmt.Sprintf("generator '%s' not found", name)))
		return
	}

	if record.Builtin != 0 {
		WriteError(w, Conflict("cannot delete built-in generator"))
		return
	}

	if _, err := g.state.Sqlite.DeleteGenerator(ctx, name); err != nil {
		WriteError(w, InternalError(err))
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

func toGeneratorResponse(g *storage.GeneratorRecord) GeneratorResponse {
	var workload interface{}
	if err := json.Unmarshal([]byte(g.Workload), &workload); err != nil {
		workload = map[string]interface{}{}
	}
	return GeneratorResponse{
		Name:        g.Name,
		Description: g.Description,
		Workload:    workload,
		Rate:        g.Rate,
		Builtin:     g.Builtin != 0,
		CreatedAt:   g.CreatedAt,
	}
}
